Sprint 11

pavelseleznev opened the pull request:
Implemented: ProgressHUD, SwiftKeychainWrapper, removed race conditions for network stack, added profile data loading and app icon, partial refactoring of storyboard into code

NaR0RoW commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -15,21 +16,47 @@ final class AuthViewController: UIViewController {
     private let oauth2Service = OAuth2Service.shared
-    private let tokenStorage = OAuth2TokenStorage()
+    private var tokenStorage = OAuth2TokenStorage()
+    private weak var loginButton: UIButton!
```
Лучше не использовать `force` операции

NaR0RoW commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -15,21 +16,47 @@ final class AuthViewController: UIViewController {
     private let showWebViewSegueIdentifier: String = "ShowWebView"
     private let oauth2Service = OAuth2Service.shared
-    private let tokenStorage = OAuth2TokenStorage()
+    private var tokenStorage = OAuth2TokenStorage()
```
`OAuth2TokenStorage` можно сделать синглтоном

NaR0RoW commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -38,19 +65,38 @@ extension AuthViewController: WebViewViewControllerDelegate {
+        let action = UIAlertAction(title: "Ok", style: .default)
+        
+        authErrorAlert.addAction(action)
+        authErrorAlert.view.accessibilityIdentifier = "authErrorAlert"
```
Было бы здорово получать названия `accessibilityIdentifier` из одного места. Например, в этом месте название индификатора могут изменить, а в тестах забудут, тогда тест упадет. Поэтому лучше названия индификаторов из одной сущности брать (например сделать структуру с статическими полями)

NaR0RoW commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -10,17 +10,50 @@ import WebKit
-    // MARK: - IB Outlets
-    @IBOutlet private weak var webView: WKWebView!
-    @IBOutlet private weak var progressView: UIProgressView!
+    lazy var webView: WKWebView = {
```
Для достижения [инкапсуляции](https://habr.com/ru/articles/87205/) лучше пометить это поле `private`

NaR0RoW commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -10,17 +10,50 @@ import WebKit
     
-    // MARK: - Internal Property
-    weak var delegate: WebViewViewControllerDelegate?
+    lazy var progressView: UIProgressView = {
```
Лучше пометить это поле `private`.
Перепроверь все свойства и методы в классах. Они должны быть максимально инкапсулированы и скрыты внутри класса

NaR0RoW commented on ImageFeed/Models/AppColor.swift:
```diff
@@ -0,0 +1,14 @@
+ 
+import UIKit
+
+struct AppColor {
+    static let ypBlack = UIColor(named: "YP Black (iOS)")
+    static let ypGray = UIColor(named: "YP Gray (iOS)")
+    static let ypWhite = UIColor(named: "YP White (iOS)")
+}
```
Отличное решение 🔥

NaR0RoW commented on ImageFeed/Models/UIBlockingProgressHUD.swift:
```diff
@@ -0,0 +1,26 @@
+///Loading indicator for time-consuming tasks (e.g. fetchOAuthToken/fetchProfile)
+final class UIBlockingProgressHUD {
+    private static var window: UIWindow? {
+        return UIApplication.shared.windows.first
```
В однострочных выражениях `return` можно опускать

NaR0RoW commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -112,4 +112,37 @@ final class ProfileViewController: UIViewController {
+        avatarImage.layer.masksToBounds = true
+        view.addSubview(avatarImage)
+        
+        avatarImage.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 32).isActive = true
+        avatarImage.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16).isActive = true
+        avatarImage.heightAnchor.constraint(equalToConstant: 70).isActive = true
+        avatarImage.widthAnchor.constraint(equalToConstant: 70).isActive = true
```
Можно таким образом констрейнты настраивать:
``` swift
NSLayoutConstraint.activate([
   avatarImage.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 32),
   avatarImage.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16),
   avatarImage.heightAnchor.constraint(equalToConstant: 70),
   avatarImage.widthAnchor.constraint(equalToConstant: 70)
])
```

NaR0RoW commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -112,4 +112,37 @@ final class ProfileViewController: UIViewController {
+        avatarImage.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16).isActive = true
+        avatarImage.heightAnchor.constraint(equalToConstant: 70).isActive = true
+        avatarImage.widthAnchor.constraint(equalToConstant: 70).isActive = true
+        self.avatarImageView = avatarImage
```
`self` везде лучше опускать, если явно того не требуется

NaR0RoW commented on ImageFeed/SceneDelegate.swift:
```diff
@@ -11,8 +11,11 @@ final class SceneDelegate: UIResponder, UIWindowSceneDelegate {
-    func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
-        guard let _ = (scene as? UIWindowScene) else { return }
+    func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectingOptions: UIScene.ConnectionOptions) {
+        guard let scene = (scene as? UIWindowScene) else { return }
```
В `Swift` лучше не использовать круглые скобки, если в них нет явной необходимости

NaR0RoW commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -8,54 +8,66 @@
+            }
+            
+            var request = URLRequest(url: url)
+            request.httpMethod = "POST"
```
Можно создать отдельный `enum` типа `String` для http методов

NaR0RoW commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2TokenStorage.swift:
```diff
@@ -6,20 +6,20 @@
-    private let userDefaults = UserDefaults.standard
+    private let keyChain = KeychainWrapper.standard
     
     private enum Keys: String {
```
В `Swift` принято `enum` в единственном числе называть, т.е. `Key`

NaR0RoW commented on ImageFeed/ServiceAccessLayer/ProfileService/ProfileImageService.swift:
```diff
@@ -0,0 +1,66 @@
+final class ProfileImageService {
+    private weak var urlSession = URLSession.shared
+    private weak var task: URLSessionTask?
+    private (set) var avatarURL: String?
```
Лишний пробел между `private` и `(set)`: `private(set)`. В `Swift 6` это будет ошибкой

NaR0RoW commented on ImageFeed/ServiceAccessLayer/ProfileService/ProfileService.swift:
```diff
@@ -0,0 +1,61 @@
+    private let oauth2TokenStorage: OAuth2TokenStorageProtocol = OAuth2TokenStorage()
+    private weak var urlSession = URLSession.shared
+    private weak var task: URLSessionTask?
+    private (set) var profile: Profile?
```
Тут тоже лишний пробел

NaR0RoW commented on ImageFeed/ServiceAccessLayer/ProfileService/ProfileService.swift:
```diff
@@ -0,0 +1,61 @@
+    private (set) var profile: Profile?
+    private init() {}
+    
+    private func fetchProfileURL(authToken: String) -> URLRequest? {
```
В целом, код читается как текст в книге, поэтому лучше вызываемые методы (`fetchProfileURL`) объявлять после методов, которые их вызывают (`fetchProfile`)

NaR0RoW commented:
Не получается авторизоваться в приложении:
```
[objectTask]: Decoding error - The data couldn’t be read because it is missing., Data: {"id":"13uQDFGbXII","updated_at":"2025-02-23T16:14:05Z","username":"nar_row","name":"Matvei Baranov","first_name":"Matvei","last_name":"Baranov","twitter_username":null,"portfolio_url":null,"bio":null,"location":null,"links":{"self":"https://api.unsplash.com/users/nar_row","html":"https://unsplash.com/@nar_row","photos":"https://api.unsplash.com/users/nar_row/photos","likes":"https://api.unsplash.com/users/nar_row/likes","portfolio":"https://api.unsplash.com/users/nar_row/portfolio","following":"https://api.unsplash.com/users/nar_row/following","followers":"https://api.unsplash.com/users/nar_row/followers"},"profile_image":{"small":"https://images.unsplash.com/profile-1739999780243-119e322fee2bimage?ixlib=rb-4.0.3\u0026crop=faces\u0026fit=crop\u0026w=32\u0026h=32","medium":"https://images.unsplash.com/profile-1739999780243-119e322fee2bimage?ixlib=rb-4.0.3\u0026crop=faces\u0026fit=crop\u0026w=64\u0026h=64","large":"https://images.unsplash.com/profile-1739999780243-119e322fee2bimage?ixlib=rb-4.0.3\u0026crop=faces\u0026fit=crop\u0026w=128\u0026h=128"},"instagram_username":null,"total_collections":0,"total_likes":22,"total_photos":0,"total_promoted_photos":0,"total_illustrations":0,"total_promoted_illustrations":0,"accepted_tos":false,"for_hire":false,"social":{"instagram_username":null,"portfolio_url":null,"twitter_username":null,"paypal_email":null},"followed_by_user":false,"photos":[],"badge":null,"tags":{"custom":[{"type":"search","title":"christmas"},{"type":"search","title":"nature"},{"type":"search","title":"outdoor"}],"aggregated":[{"type":"search","title":"3d"},{"type":"search","title":"accessory"},{"type":"search","title":"art"},{"type":"search","title":"christmas"},{"type":"search","title":"craft"},{"type":"search","title":"cutout"},{"type":"search","title":"decor"},{"type":"search","title":"decoration"},{"type":"search","title":"digital image"},{"type":"search","title":"electronic"},{"type":"search","title":"face"},{"type":"search","title":"head"},{"type":"search","title":"holiday"},{"type":"search","title":"human"},{"type":"search","title":"jewelry"},{"type":"search","title":"modern"},{"type":"search","title":"nature"},{"type":"search","title":"necklace"},{"type":"search","title":"outdoor"},{"type":"search","title":"paper"},{"type":"search","title":"phone"},{"type":"search","title":"photo"},{"type":"search","title":"photography"},{"type":"search","title":"portable ssd"},{"type":"search","title":"project"},{"type":"search","title":"resource database™"},{"type":"search","title":"selfie"},{"type":"search","title":"snowflake"},{"type":"search","title":"storage"},{"type":"search","title":"video camera"}]},"followers_count":0,"following_count":0,"allow_messages":true,"numeric_id":17616078,"downloads":0,"meta":{"index":false},"uid":"13uQDFGbXII","confirmed":true,"uploads_remaining":10,"unlimited_uploads":false,"email":"nar0row@gmail.com","dmca_verification":"unverified","unread_in_app_notifications":false,"unread_highlight_notifications":false}
[objectTask]: URLSession error - Failed to load profile data: The data couldn’t be read because it is missing.
[fetchOAuthToken]: Error profileData - The data couldn’t be read because it is missing.
```

pavelseleznev commented:
> Не получается авторизоваться в приложении:

Матвей, спасибо большое за ревью. Постарался внести все изменения.

pavelseleznev commented:
Applied critical/non-critical post review fixes:

modified objectTask / fetchOAuthToken behavior to handle cases where profile may have missing data;
modified color scheme;
removed force-unwrap usage;
other code changes.

NaR0RoW commented on ImageFeed/Models/HTTPMethod.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+enum HTTPMethod {
+    static let typePost: String = "POST"
+}
```
Я бы таким образом `enum` объявил:
``` swift
enum HTTPMethod: String {
    case post = "POST"
    case get = "GET"
    case delete = "DELETE"
}
```

Использование:
``` swift
request.httpMethod = HTTPMethod.post.rawValue
```

NaR0RoW commented on ImageFeed/Models/AuthErrorAlert.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+enum AuthErrorAlert {
+    static let accessibilityAlertIdentifier: String = "authErrorAlert"
+}
```
Я бы наверное таким образом сущность для `accessibilityIdentifiers` организовал:
``` swift
struct AccessibilityIdentifiers {
    static let authErrorAlert: String = "authErrorAlert"
}
```
или
``` swift
enum AccessibilityIdentifier: String {
    case authErrorAlert
}
```

NaR0RoW reviewed (approved):
Отличная работа 🔥

pavelseleznev commented:
> Отличная работа 🔥

Спасибо ещё раз.

pavelseleznev commented:
Post review changes:
added non-critical fixes.
